### PR TITLE
RN-699: Enforce permissions checking on both Data Elements and Entities in data-broker

### PR DESCRIPTION
**Pull request by rohan-bes**
### Issue RN-699:

So it turns out that `data-broker` has just been checking if the user has access to the requested data-elements permission groups anywhere, instead of checking that they have access to the permission groups in the entities they're requesting data for. This was never really the intended behaviour, so it's possible this got lost in merge conflicts or something...

Anyway, this PR is to enforce the stricter permissions check against data-elements and entities.

### CODE REVIEW

File: packages/access-policy/src/AccessPolicy.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+export class AccessPolicy {
```
Comment (rohan-bes) [inline packages/access-policy/src/AccessPolicy.js:6]: This is an absolutely superficial change. `AccessPolicy` has always been one of the very rare cases where we keep a class in a file named something other than the class name. I find myself constantly forgetting this and struggling to find it when searching in vscode, so I moved it. Hope the reviewer agrees and sorry for doing it here but I didn't feel it warranted a PR of its own 😅

File: packages/access-policy/src/index.js [hunk 2]
```diff
@@ -3,130 +3,4 @@
...
-   *
-   * @returns entities[] The entity objects
-   */
-  getEntitiesAllowed(permissionGroup) {
-    const allEntityCodes = Object.keys(this.policy);
-    if (!permissionGroup) return allEntityCodes;
-    return allEntityCodes.filter(e => this.allows(e, permissionGroup));
-  }
-}
+export { AccessPolicy } from './AccessPolicy';
```
Comment (rohan-bes) [inline packages/access-policy/src/index.js:6]: Moved to `AccessPolicy.js`

File: packages/data-broker/src/DataBroker.js [hunk 3]
```diff
@@ -85,33 +88,66 @@ export class DataBroker {
...
-      ) {
-        continue;
+
+    const getDataElementsWithMissingPermissions = permissions =>
+      dataElements
+        .filter(element => element.permission_groups.length > 0)
+        .filter(element => !element.permission_groups.some(group => permissions.includes(group)))
+        .map(element => element.code);
+
+    if (countryCodes === null) {
```
Comment (rohan-bes) [inline packages/data-broker/src/DataBroker.js:112]: It's possible for countryCode to be null when fetching for `sync_groups` I believe. At the moment sync groups don't do permissions checks, but they could... so leaving this here for now 👍

File: packages/data-broker/src/DataBroker.ts [hunk 4]
```diff
@@ -144,33 +153,70 @@ export class DataBroker {
...
-      ) {
-        continue;
+
+    const getDataElementsWithMissingPermissions = (permissions: string[]) =>
+      (dataElements as DataElement[])
+        .filter(element => element.permission_groups.length > 0)
+        .filter(element => !element.permission_groups.some(group => permissions.includes(group)))
+        .map(element => element.code);
+
+    if (countryCodes === null) {
```
Comment (IgorNadj) [inline packages/data-broker/src/DataBroker.ts:178]: I'd love to drop support for no country codes if possible, I can't wrap my head around it. In fact I don't know if it's ever done?

File: packages/data-broker/src/__tests__/DataBroker/DataBroker.test.ts [hunk 5]
```diff
@@ -329,6 +330,88 @@ describe('DataBroker', () => {
...
+              type: 'dataElement',
+            },
+            { organisationUnitCodes: ['TO'] },
+          ),
+        ).toBeRejectedWith(
+          `Missing permissions to the following data elements:\nTO: RESTRICTED_01`,
+        );
+      });
+
+      it("throws an error if any of entities in fetch the user doesn't have required permissions for", async () => {
```
Comment (IgorNadj) [inline packages/data-broker/src/__tests__/DataBroker/DataBroker.test.ts:374]: Beautiful tests

**Review by IgorNadj — APPROVED** [review]
Looks great nice one! Thanks

File: packages/data-broker/src/DataBroker.ts (hunk 4 above)
Comment (IgorNadj) [reply packages/data-broker/src/DataBroker.ts:178]: Ah just saw your other comment, all good.

**Pull request merged**